Core/Character.cs

```diff
@@ -1,9 +1,128 @@
+
+        public CharacterVariable GetVariable(string name)
+        {
+            if (Variables.TryGetValue(name, out var result))
+                return result;
+            result = new CharacterVariable(this, new List<GameValue>());
+            Variables.Add(name, result);
+            return result;
+        }
+
+        public BonusBuilder SetVariableBase(string name, int value)
+        {
+            var fixedValue = new FixedValue("BASE", value);
+            if (Variables.TryGetValue(name, out CharacterVariable stat))
+            {
+                stat.Replace("BASE", fixedValue);
+            }
+            else
+            {
+                Variables.Add(name, new CharacterVariable(this, fixedValue));
+            }
+
+            return new BonusBuilder(this, name);
+        }
+
+        public BonusBuilder AddToVariable(string name, string type, int value)
+        {
+            GetVariable(name).Add(new FixedValue(type, value));
+            return new BonusBuilder(this, name);
+        }
+
+        public BonusBuilder AddToVariable(string name, int value)
+        {
+            return AddToVariable(name, name, value);
+        }
+
+        public BonusBuilder AddReferenceToVariable(string name, string type, string referenceStatistic)
+        {
+            GetVariable(name).Add(new ComputedValue(type, GetVariable(referenceStatistic)));
+            return new BonusBuilder(this, name);
+        }
+
+        public BonusBuilder AddReferenceToVariable(string name, string referenceStatistic)
+        {
+            return AddReferenceToVariable(name, name, referenceStatistic);
+        }
+
+        public BonusBuilder DefineVariable(string name)
+        {
+            // This doesn't need to do anything, because "GetReference" will always create a variable that doesn't exist
+            // and set it to zero.  It's just a convenience to callers to get a BonusBuilder
+            return new BonusBuilder(this, name);
+        }
+
+        public BonusBuilder AddReferenceToVariable(string name, string type, string referenceStatistic, ComputeValue computation)
+        {
+            GetVariable(name).Add(new ComputedValue(type, GetVariable(referenceStatistic), computation));
+            return new BonusBuilder(this, name);
+        }
+
+        public BonusBuilder AddReferenceToVariable(string name, string referenceStatistic, ComputeValue computation)
+        {
+            return AddReferenceToVariable(name, name, referenceStatistic, computation);
+        }
+
+        public struct BonusBuilder
+        {
+            private readonly Character _character;
+            private readonly string _name;
+
+            public BonusBuilder(Character character, string name)
+            {
+                _character = character;
+                _name = name;
+            }
+
+            public BonusBuilder Add(int value)
+            {
+                return Add(_name, value);
```

Comment:
Why is this passing `_name` for type?

Reply:
It was trying to encode the weird rule that unnamed things from the same source don't stack... (Like if you get an untyped bonus equal to your STR score from somewhere, and another untyped bonus equal to your strength score, those don't stack).

But maybe it should just leave them "untyped" and leave that piece of logic for later.

Reply:
Changed it to the empty string.  We can deal with the weird stacking rules later somehow. :-)

Reply:
Yea, I think either "null" or "empty string" is enough to encode "untyped" bonuses

Reply:
Switched it to the empty string, we'll let the rule engine handle the other part somewhere else. :-) I was trying to be too clever.